### ISSUE TITLE
kld(q(zk|x)||p(zk)) learns to be lower than 0

### ISSUE THREAD

**Issue by limingzhan0904** [siddsax/Normalizing_Flows#1]
hi, Siddhartha, when you reimplement NFs, did the q(zk|x)||p(zk) turn out to be lower than 0  when flows number increased(such as 80 or bigger)?

**Comment by siddsax** [comment]
Hi, KL part here is calculated with the KL between unit gaussian whereas a log-determinant term calculates the free energy for the flows ( which is negative ). The former is always positive by formulae. `0.5 * tf.reduce_sum(tf.exp(z_logvar) + z_mu**2 - 1. - z_logvar, 1)`. So ideally there shouldn't be an issue like that. Though we didn't check for that many numbers of flows. 
On an aside I noticed that the experts were fixed to 4 in the code, so I have updated it now.

**Comment by limingzhan0904** [comment]
Actually, I think the former we shouldn't use the analytical formula to calculate, 'cause we can not assume that zk =z0. But anyway, I do observe that the kld will be negative if flows grow.

**Comment by siddsax** [comment]
This was some time back. Although what I remember is that the elbo for VAE+Flows actually reduces to that of VAE minus the free energy term (Eq 15 in their paper, expectation wrt q0. This is the good thing about the method as one can do VI with a simple gaussian but have a complex posterior). 
I also re-ran the models for some iterations and I am not getting a negative KL and the empirical formulae to calculate is by definition positive, plot (e^x - x -1). Weird.
